source/nodejs/adaptivecards/src/card-elements.ts

```diff
@@ -6579,6 +6581,27 @@ export class AdaptiveCard extends ContainerWithActions {
                 });
         }
 
+        let allInputs = this.getAllInputs();
```

Comment:
>getAllInputs [](start = 29, length = 12)

The other platforms are enforcing ID uniqueness for all elements, not just inputs. Also, there is logic for allowing ids to conflict within a given fallback hierarchy but not across it. @paulcam206

Reply:
The full description of the algorithm with justification can be found [here](https://github.com/microsoft/AdaptiveCards/blob/master/source/shared/cpp/ObjectModel/ParseContext.cpp#L26).

---
In reply to: [283443447](https://github.com/microsoft/AdaptiveCards/pull/2905#discussion_r283443447) [](ancestors = 283443447)

Reply:
The JS renderer has always separated parse-time and post-parse validation precisely for these kinds of things.

Fallback is logic that happens at parse time. As the JS renderer parses a payload, it builds an object model representation of the parsed card; for any given element, either that element or its fallback ends up in the OM. At that point no Id validation needs to be applied. Then, once the full OM is generated, a second validation pass can be performed to validate Ids, among other things.

Splitting validation into parse-time validation and post-parse validation allows for much easier Id validation logic and plenty of other things, such as, for instance, easily verifying that a ToggleVisibility action references Ids that are actually used in the card.

Reply:
If the JS validation is strictly post-parse, then I think we're *close* to fine. Consider this card snip, though:

```json
{
    "type": "Input.AdaptiveCard2.0.Feature",
    "fallback": {
        "type": "Input.Text",
        "id":   "uhoh"
    }
},
{
    "type": "Input.Text",
    "id":   "uhoh"
}
```

So you're using your 2.0 input element and the 2.0 designer/renderer to test. Everything looks good. The card validates fine because fallback was never invoked. You take your validated card and drop it on a 1.2 renderer and you'll hit an error because now you've got a conflict.

Reply:
That's an interesting point. How does your logic handle the below payload? From what I read it seems it will pass validation but I might be wrong:

```
{
	"type": "Container",
	"items": [
		{
			"type": "Adaptive.2.0.Feature",
			"fallback": {
				"type": "TextBlock",
				"id": "fallback",
				"text": "Nice"
			}
		},
		{
			"type": "Adaptive.2.1.Feature",
			"fallback": {
				"type": "TextBlock",
				"id": "fallback",
				"text": "Nice"
			}
		}
	]
}
```

Reply:
The point I want to make is, for the design time experience to actually make it easy for an author to know what will work and what won't, in my opinion you need to include a version selector in the designer, while in "Preview mode." As the user selects a version, the same payload is parsed and validated as it would be on a host that only supports that version. This allows for accurate warnings/errors and also to actually test fallback. And if that is what we end up doing, I am not sure we need that complex parse-time Id validation logic.

Reply:
The above card in UWP visualizer:

![image](https://user-images.githubusercontent.com/16614499/57728666-4db66500-7649-11e9-97b6-f310a58ab515.png)

It's probably worth noting that instead of "Some New Adaptive Cards Feature", it could easily be a custom element that the card author may have implemented in some but not all renderers...

Reply:
Nice! At this point, I will not be implementing this algorithm in the JS renderer because it would require slightly changing the overall way payloads are parsed, and would most likely introduce a breaking change.